ftl/php/builder.py

```diff
@@ -18,63 +18,110 @@
+            self._destination_path = destination_path
+
+        def GetCacheKey(self):
+            descriptor_contents = ftl_util.descriptor_parser(
+                self._descriptor_files, self._ctx)
+
+            return hashlib.sha256(descriptor_contents).hexdigest()
+
+        def BuildLayer(self):
+            """Override."""
+            overrides = self._generate_overrides()
+
+            layer, sha = self._gen_package_tar(self._pkg_txt,
+                                               self._destination_path)
+            logging.info('Generated layer with sha: %s', sha)
+
+            return layer, sha, overrides
+
+        def _gen_package_tar(self, pkg_txt, destination_path):
+            # Create temp directory to write package descriptor to
+            pkg_dir = tempfile.mkdtemp()
+            app_dir = os.path.join(pkg_dir, destination_path.strip("/"))
+            os.makedirs(app_dir)
+
+            # Copy out the relevant package descriptors to a tempdir.
+            ftl_util.descriptor_copy(self._ctx, self._descriptor_files,
+                                     app_dir)
 
-    def CreatePackageBase(self, base, destination_path="/workspace/app"):
-        """Override."""
-        overrides = self._generate_overrides()
-        layer, sha = self._gen_package_tar(destination_path)
-        logging.info('Generated layer with sha: %s', sha)
-
-        with append.Layer(
-                base, layer, diff_id=sha, overrides=overrides) as dep_image:
-            return dep_image
-
-    def _gen_package_tar(self, destination_path):
-        # Create temp directory to write package descriptor to
-        pkg_dir = tempfile.mkdtemp()
-        app_dir = os.path.join(pkg_dir, destination_path.strip("/"))
-        os.makedirs(app_dir)
-
-        # Copy out the relevant package descriptors to a tempdir.
-        for f in self.descriptor_files:
-            if self._ctx.Contains(f):
-                with open(os.path.join(app_dir, f), 'w') as w:
-                    w.write(self._ctx.GetFile(f))
-
-        subprocess.check_call(['rm', '-rf', os.path.join(app_dir, 'vendor')])
-
-        with ftl_util.Timing("composer_install"):
             subprocess.check_call(
-                ['composer', 'install', '--no-dev', '--no-scripts'],
-                cwd=app_dir)
-
-        return ftl_util.zip_dir_to_layer_sha(pkg_dir)
-
-
-def From(ctx):
-    return PHP(ctx)
+                ['rm', '-rf', os.path.join(app_dir, 'vendor')])
+
+            with ftl_util.Timing("composer_install"):
+                if pkg_txt is None:
+                    try:
+                        subprocess.check_call(
+                            [
+                                'composer', 'install', '--no-dev',
+                                '--no-scripts'
+                            ],
+                            cwd=app_dir)
+                    except subprocess.CalledProcessError as e:
+                        print e.output
+                        exit(1)
```

Comment:
Drop the print

Reply:
Also handle this exception a little more cleanly. Can you catch it higher up?

Reply:
I'm just going to remove the try/except here